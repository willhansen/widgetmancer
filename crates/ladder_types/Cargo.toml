[package]
name = "ladder_types"
version = "0.0.0"
edition.workspace = true

[dependencies]
typenum = "1.17.0"
trybuild = "1.0"
num.workspace = true
derive_more.workspace = true
