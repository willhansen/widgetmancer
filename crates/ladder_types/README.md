A wrapper-type for the phenomenon of subtracting two points resulting in a vector rather than a point.

This crate differs from other approaches (eg euclid) in that rather than the difference of two vectors being a vector, it instead reaches a third level, and so on.

Statically checked

`ThingWithRelativity<Contents, Level=0> ==> T<0>`

`T<0> - T<0> ==> T<1>`
`T<1> - T<1> ==> T<2>`
`T<N> - T<N> ==> T<N+1>`

`T<0> +/- T<1> ==> T<0>`
`T<N> +/- T<N+1> ==> T<0>`

`T<N> + T<M>, where N!=M ==> INVALID`
`T<N> - T<M>, where N!=M and N!=M+1 ==> INVALID`

May have another static parameter for allowing opt-in behavior.  eg allowing `ThingWithRelativity<i32, Level=0>::new(5) + 7`
- Assume no relativity implies level of 0, or whatever the other level is (or one higher?)?
- Maybe an enum for default relativity level?  Options: Same, PlusOne, Zero, Disallow
