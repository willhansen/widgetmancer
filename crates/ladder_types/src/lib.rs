pub mod kiss_ladder;
pub mod typenum_ladder;
pub mod unary_wrapper_ladder;
pub mod numbers_on_ladders;


