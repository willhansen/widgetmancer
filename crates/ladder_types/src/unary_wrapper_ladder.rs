use std::ops::{Add, Sub};
use std::marker::PhantomData;

pub trait LadderRung {
    fn new() -> Self;
}

#[derive(Eq, PartialEq, Debug)]
pub struct RelativeTo<T>(PhantomData<T>);


impl<T: LadderRung> LadderRung for RelativeTo<T> { 
    fn new() -> Self {
        Self(PhantomData::<T>)
    }
}
impl LadderRung for RelativeTo<()> { 
    fn new() -> Self {
        Self(PhantomData::<()>)
    }
}

pub type R0 = RelativeTo<()>;
pub type R1 = RelativeTo<R0>;
pub type R2 = RelativeTo<R1>;
pub type R3 = RelativeTo<R2>;
pub type R4 = RelativeTo<R3>;
// ...


impl<R> Add<RelativeTo<RelativeTo<R>>> for RelativeTo<R>
    where
    RelativeTo<R>: LadderRung
{
    type Output = RelativeTo<R>;

    fn add(self, _rhs: RelativeTo<RelativeTo<R>>) -> Self::Output {
        Self::Output::new()
    }
} 

impl<R> Sub<RelativeTo<RelativeTo<R>>> for RelativeTo<R>
    where
    RelativeTo<R>: LadderRung
{
    type Output = RelativeTo<R>;

    fn sub(self, _rhs: RelativeTo<RelativeTo<R>>) -> Self::Output {
        Self::Output::new()
    }
} 

impl<R> Sub<RelativeTo<R>> for RelativeTo<R>
    where
    RelativeTo<R>: LadderRung
{
    type Output = RelativeTo<RelativeTo<R>>;

    fn sub(self, _rhs: RelativeTo<R>) -> Self::Output {
        Self::Output::new()
    }
} 

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_add() {
        assert_eq!(R0::new() + R1::new(), R0::new());
        assert_eq!(R1::new() + R2::new(), R1::new());
    }
    #[test]
    fn test_sub() {
        assert_eq!(R0::new() - R1::new(), R0::new());
        assert_eq!(R1::new() - R2::new(), R1::new());

        assert_eq!(R0::new() - R0::new(), R1::new());
        assert_eq!(R1::new() - R1::new(), R2::new());
    }
}
