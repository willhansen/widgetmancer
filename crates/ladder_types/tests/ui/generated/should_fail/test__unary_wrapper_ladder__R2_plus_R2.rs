
use ladder_types::unary_wrapper_ladder::*;

fn main() {
   let _ = R2::new() + R2::new();
}
