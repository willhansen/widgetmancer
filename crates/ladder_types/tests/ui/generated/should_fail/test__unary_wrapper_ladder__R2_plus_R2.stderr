error[E0308]: mismatched types
 --> tests/ui/generated/should_fail/test__unary_wrapper_ladder__R2_plus_R2.rs:5:24
  |
5 |    let _ = R2::new() + R2::new();
  |                        ^^^^^^^^^ expected `RelativeTo<()>`, found `()`
  |
  = note: expected struct `RelativeTo<RelativeTo<RelativeTo<RelativeTo<()>>>>`
             found struct `RelativeTo<RelativeTo<RelativeTo<()>>>`
